### PR TITLE
Terminate connections correctly when calling close

### LINKED ISSUE #606: mysql2 gem improperly closing MySQL connections
The mysql2 gem doesn't properly close the connection, even when explicitly calling the 'client.close' method, resulting in increase of 'Aborted_clients' counter indicating an improper closure of the connection. My setup is MySQL 5.6.23 installed from MySQL's ubuntu repo on an Ubuntu Trusty EC2 instance, mysql2 gem v0.3.18, ruby 2.0.0p353 (embedded sensu version), libmysqlclient18

```
dtest@ip-10-1-1-73:~# /opt/sensu/embedded/bin/ruby --version
ruby 2.0.0p353 (2013-11-22 revision 43784) [x86_64-linux]
dtest@ip-10-1-1-73:~# /opt/sensu/embedded/bin/gem list | grep mysql
mysql (2.9.1)
mysql2 (0.3.18)
dtest@ip-10-1-1-73:~# ldd /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/mysql2-0.3.18/lib/mysql2/mysql2.so
  linux-vdso.so.1 =>  (0x00007fffa07fe000)
  libruby.so.2.0 => /opt/sensu/embedded/lib/libruby.so.2.0 (0x00007f8bde4ec000)
  libmysqlclient.so.18 => /usr/lib/x86_64-linux-gnu/libmysqlclient.so.18 (0x00007f8bddf35000)
  libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8bddd10000)
  libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8bdd94b000)
  librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f8bdd743000)
  libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f8bdd53e000)
  libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007f8bdd305000)
  libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8bdcfff000)
  /lib64/ld-linux-x86-64.so.2 (0x00007f8bdeba9000)
  libz.so.1 => /opt/sensu/embedded/lib/libz.so.1 (0x00007f8bdcde1000)
  libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f8bdcadd000)
  libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f8bdc8c7000)
dtest@ip-10-1-1-73:~# mysql --version
mysql  Ver 14.14 Distrib 5.6.23, for Linux (x86_64) using  EditLine wrapper
```

I have the following example script:
```ruby
#!/usr/bin/ruby

require 'rubygems' if RUBY_VERSION < '1.9.0'
require 'mysql2'

puts Time.now.strftime("%d/%m/%Y %H:%M:%S")+"-Before connect";
# socket
client = Mysql2::Client.new(:host => 'localhost', :username => 'ruby', :password => 'test', :socket => "/var/lib/mysql/mysql.sock");
# tcp
#client = Mysql2::Client.new(:host => '127.0.0.1', :username => 'ruby', :password => 'test');
puts Time.now.strftime("%d/%m/%Y %H:%M:%S")+"- Query 1";
client.query("SELECT 1;");
sleep(1);
puts Time.now.strftime("%d/%m/%Y %H:%M:%S")+"- Query 2";
client.query("SELECT 1;");
sleep(1);
puts Time.now.strftime("%d/%m/%Y %H:%M:%S")+"- Query 3";
client.query("SELECT 1;");
sleep(1);
puts Time.now.strftime("%d/%m/%Y %H:%M:%S")+"- Close";
client.close;
sleep(1);
puts Time.now.strftime("%d/%m/%Y %H:%M:%S")+"- End";

```

Which produces output like this, for both socket and TCP connections:
```
dtest@ip-10-1-1-73:~# mysql --defaults-file=~/.my.cnf -e "SHOW GLOBAL STATUS LIKE 'aborted_clients'";/opt/sensu/embedded/bin/ruby mysql_test.mysql2.tcp.rb ;mysql --defaults-file=~/.my.cnf -e "SHOW GLOBAL STATUS LIKE 'aborted_clients'";
+-----------------+-------+
| Variable_name   | Value |
+-----------------+-------+
| Aborted_clients | 12106 |
+-----------------+-------+
02/04/2015 00:00:32-Before connect
02/04/2015 00:00:32- Query 1
02/04/2015 00:00:33- Query 2
02/04/2015 00:00:34- Query 3
02/04/2015 00:00:35- Close
02/04/2015 00:00:36- End
+-----------------+-------+
| Variable_name   | Value |
+-----------------+-------+
| Aborted_clients | 12107 |
+-----------------+-------+
dtest@ip-10-1-1-73:~# mysql --defaults-file=~/.my.cnf -e "SHOW GLOBAL STATUS LIKE 'aborted_clients'";/opt/sensu/embedded/bin/ruby mysql_test.mysql2.socket.rb ;mysql --defaults-file=~/.my.cnf -e "SHOW GLOBAL STATUS LIKE 'aborted_clients'";
+-----------------+-------+
| Variable_name   | Value |
+-----------------+-------+
| Aborted_clients | 12109 |
+-----------------+-------+
02/04/2015 00:01:26-Before connect
02/04/2015 00:01:26- Query 1
02/04/2015 00:01:27- Query 2
02/04/2015 00:01:28- Query 3
02/04/2015 00:01:29- Close
02/04/2015 00:01:30- End
+-----------------+-------+
| Variable_name   | Value |
+-----------------+-------+
| Aborted_clients | 12110 |
+-----------------+-------+
dtest@ip-10-1-1-73:~# grep ruby /var/log/mysql/error.log
2015-04-02 00:00:35 11797 [Warning] Aborted connection 12561 to db: 'unconnected' user: 'ruby' host: '127.0.0.1' (Got an error reading communication packets)
2015-04-02 00:01:29 11797 [Warning] Aborted connection 12566 to db: 'unconnected' user: 'ruby' host: 'localhost' (Got an error reading communication packets)
```
We can see the entry in the error log at the same time the script calls the 'client.close' method.

I have verified that the mysql gem does not increase the 'Aborted_clients' counter with the same script and linked libraries.

### PR DESCRIPTION
**Pull request by cbandy**
Fixes #606

### CODE REVIEW

File: ext/mysql2/client.c [hunk 1]
```diff
@@ -214,24 +214,8 @@ static void *nogvl_close(void *ptr) {
...
-      fprintf(stderr, "[WARN] mysql2 failed to invalidate FD safely, leaking some memory\n");
-      close(wrapper->client->net.fd);
-      return NULL;
-    }
-#endif
-
-    mysql_close(wrapper->client); /* only used to free memory at this point */
+    mysql_close(wrapper->client);
   }
-
```
Comment (tamird) [inline ext/mysql2/client.c]: spurious whitespace change

File: ext/mysql2/client.c [hunk 2]
```diff
@@ -245,6 +229,22 @@ void decr_mysql2_client(mysql_client_wrapper *wrapper)
   wrapper->refcount--;
 
   if (wrapper->refcount == 0) {
+    if (wrapper->connected) {
```
Comment (tamird) [inline ext/mysql2/client.c]: this block of code needs a comment explaining why it's not in `nogvl_close`

Comment (tamird) [inline ext/mysql2/client.c]: also, the `if (wrapper->connected) {` should be inside the `#ifndef`

**Comment by sodabrew** [comment]
+1 for incorporating the comments from @tamird and please rebase. I'll merge this week.

**Comment by cbandy** [comment]
> this block of code needs a comment explaining why it's not in `nogvl_close`

I've changed the wording slightly to emphasize that it being called during garbage collection.

**Comment by cbandy** [comment]
There is only one other call to `invalidate_fd` and the comment above it has me concerned:

```c
static VALUE disconnect_and_raise(VALUE self, VALUE error) {                                                                                                                
  GET_CLIENT(self);                                                                                                                                                         
                                                                                                                                                                            
  MARK_CONN_INACTIVE(self);                                                                                                                                                 
  wrapper->connected = 0;                                                                                                                                                   
                                                                                                                                                                            
  /* Invalidate the MySQL socket to prevent further communication.                                                                                                          
   * The GC will come along later and call mysql_close to free it.                                                                                                          
   */                                                                                                                                                                       
  if (invalidate_fd(wrapper->client->net.fd) == Qfalse) {                                                                                                                   
    fprintf(stderr, "[WARN] mysql2 failed to invalidate FD safely, closing unsafely\n");                                                                                    
    close(wrapper->client->net.fd);                                                                                                                                         
  }                                                                                                                                                                         
                                                                                                                                                                            
  rb_exc_raise(error);                                                                                                                                                      
}
```

However, `mysql_close` is only called when `wrapper->connected`:

```c
static void *nogvl_close(void *ptr) {                                                                                                                                       
  mysql_client_wrapper *wrapper;                                                                                                                                            
  wrapper = ptr;                                                                                                                                                            
  if (wrapper->connected) {                                                                                                                                                 
    MARK_CONN_INACTIVE(self);                                                                                                                                               
    wrapper->connected = 0;                                                                                                                                                 
    mysql_close(wrapper->client);                                                                                                                                           
  }                                                                                                                                                                         
                                                                                                                                                                            
  return NULL;                                                                                                                                                              
} 
```

**Comment by sodabrew** [comment]
Thanks for the rebase, and nice find on that corner case! For sure we need to avoid leaking the connection handle.

Let's take a look at where `disconnect_and_raise` is called - maybe those are situations where we need to burn the connection for all users and so the original mysql_close behavior would be more correct anyways.

**Comment by sodabrew** [comment]
How about this:

```
diff --git a/ext/mysql2/client.c b/ext/mysql2/client.c
index 71deb99..7649234 100644
--- a/ext/mysql2/client.c
+++ b/ext/mysql2/client.c
@@ -209,12 +209,13 @@ static VALUE invalidate_fd(int clientfd)
 #endif /* _WIN32 */

 static void *nogvl_close(void *ptr) {
-  mysql_client_wrapper *wrapper;
-  wrapper = ptr;
-  if (wrapper->connected) {
-    MARK_CONN_INACTIVE(self);
-    wrapper->connected = 0;
+  mysql_client_wrapper *wrapper = ptr;
+
+  if (wrapper->client) {
     mysql_close(wrapper->client);
+    wrapper->client = NULL;
+    wrapper->connected = 0;
+    wrapper->active_thread = Qnil;
   }

   return NULL;
```

**Pull request merged**